script/tool/lib/src/version_check_command.dart

```diff
@@ -200,27 +206,73 @@ class VersionCheckCommand extends PackageLoopingCommand {
+    // PR with post release label is going to sync changelog.md and pubspec.yaml
+    // change back to main branch. We can proceed with ragular version check.
+    final bool hasPostReleaseLabel =
+        _prLabels.contains('post-release-${pubspec.name}');
+    bool versionChanged = false;
+
+    if (usesBatchRelease && !hasPostReleaseLabel) {
+      final String relativePackagePath = await _getRelativePackagePath(package);
+      final List<String> changedFilesInPackage = changedFiles
+          .where((String path) => path.startsWith(relativePackagePath))
+          .toList();
+
+      // For batch release, we only check pending changelog files.
+      final List<PendingChangelogEntry> allChangelogs;
+      try {
+        allChangelogs = package.getPendingChangelogs();
+      } on FormatException catch (e) {
+        errors.add(e.message);
+        return PackageResult.fail(errors);
+      }
+
+      final List<PendingChangelogEntry> newEntries = allChangelogs
+          .where((PendingChangelogEntry entry) => changedFilesInPackage.any(
+              (String path) => path.endsWith(entry.file.path.split('/').last)))
+          .toList();
+      versionChanged = newEntries.any(
+          (PendingChangelogEntry entry) => entry.version != VersionChange.skip);
+
+      // The changelog.md and pubspec.yaml's version should not be updated directly.
+      if (changedFilesInPackage.contains('$relativePackagePath/CHANGELOG.md')) {
+        printError(
+            'This package uses batch release, so CHANGELOG.md should not be changed directly.\n'
+            'Instead, create a pending changelog file in pending_changelogs folder.');
+        errors.add('CHANGELOG.md changed');
+      }
+      if (changedFilesInPackage.contains('$relativePackagePath/pubspec.yaml')) {
+        if (versionState != _CurrentVersionState.unchanged) {
+          printError(
+              'This package uses batch release, so the version in pubspec.yaml should not be changed directly.\n'
+              'Instead, create a pending changelog file in pending_changelogs folder.');
+          errors.add('pubspec.yaml version changed');
+        }
+      }
+    } else {
+      switch (versionState) {
+        case _CurrentVersionState.unchanged:
+          versionChanged = false;
+        case _CurrentVersionState.validIncrease:
+        case _CurrentVersionState.validRevert:
+        case _CurrentVersionState.newPackage:
+          versionChanged = true;
+        case _CurrentVersionState.invalidChange:
+          versionChanged = true;
+          errors.add('Disallowed version change.');
+        case _CurrentVersionState.unknown:
+          versionChanged = false;
+          errors.add('Unable to determine previous version.');
+      }
+
+      if (!(await _validateChangelogVersion(package,
+          pubspec: pubspec, pubspecVersionState: versionState))) {
+        errors.add('CHANGELOG.md failed validation.');
+      }
     }
 
```

Comment:
There should be a new check that a package that does not use batched release isn't adding any pending changelog files, and a clear error message pointing to docs about batched vs unbatched releases.

Since we will have two ways of doing things, we should make sure the tooling prevents people from accidentally doing the wrong thing in the wrong package, both since it will be hard for reviewers to reliably catch that kind of mistake (because it will look like other changes they are used to seeing in other packages), and because new contributors may cargo-cult from other PRs and the burden of pointing them to the right way of doing things for the current package should be on the tooling rather than humans.